core/sql/optimizer/costmethod.cpp

```diff
@@ -13997,15 +13997,188 @@ CostMethodFastExtract::computeOperatorCostInternal(RelExpr* op,
+  // inserting at the end or the beginning of the key range in a 
+  // partition; intuitively inserting in the middle would seem to incur
+  // the full insertion cost. This is worthwhile taking into 
+  // consideration as it opens the possibility of choosing between a
+  // plan that sorts rows in Trafodion before passing them to HBase
+  // vs. a plan that does not. To make this calculation we must know
+  // the memstore insertion cost (point 1 above), the order of the 
+  // probes, whether the ESPs are aligned to the Regions of the 
+  // target table, and whether we are inserting at the beginning or
+  // end of the key range. A first approximation to the last item
+  // would be whether the target table is empty. This is interesting
+  // because the case that we are doing an INSERT/SELECT into a new
+  // table is likely to be common.
+  //
+  // 3. Take into account memstore flush cost. We could add I/O time
+  // for flushes. For example, we could compare the number of probes
+  // per HBase Region with the number of rows that would cause a
+  // flush (the latter can be obtained from 
+  // HbaseClient::estimateMemStoreRows() and is a function of the
+  // HBase parameter hbase.hregion.memstore.flush.size). Again, this
+  // cost will be paid no matter the plan choice so this is not
+  // interesting today. As with point 1, it becomes interesting only
+  // if there is a plan choice between inserting via memstore or not.
+  //
+  // In the interest of time, we move forward without these 
+  // improvements for now.
+
+  cvFR.setNumProbes(noOfProbesPerStream);
+  cvLR.setNumProbes(noOfProbesPerStream);
+
+  // ---------------------------------------------------------------------
+  // Synthesize and return cost object.
+  // ---------------------------------------------------------------------
+
+  Cost *costPtr = new STMTHEAP
+    Cost(&cvFR
+         , &cvLR
+         , NULL
+         , Lng32(activeCpus_.getValue())
+         , Lng32(streamsPerCpu_.getValue())
+         );
+
+#ifndef NDEBUG
+  if (CmpCommon::getDefault(OPTIMIZER_PRINT_COST) == DF_ON)
+    {
+      pfp = stdout;
+      fprintf(pfp, "HbaseInsert elapsed time: ");
+      fprintf(pfp, "%f", costPtr->
+              convertToElapsedTime(
+                                   myContext->getReqdPhysicalProperty()).
+              value());
+      fprintf(pfp, "\n");
+    }
+#endif
+
+  return costPtr;
+} // CostMethodHbaseInsert::computeOperatorCostInternal
+
+// -----------------------------------------------------------------------
+// CostMethodHbaseInsert::cleanUp()
+//
+// The method cleans up cached parameters which need deallocation and
+// should be called after a costing session is done.
+// -----------------------------------------------------------------------
+void CostMethodHbaseInsert::cleanUp()
+{
+  activePartitions_ = csOne;
+  activeCpus_ = csOne;
+  streamsPerCpu_ = csOne;
+  countOfAsynchronousStreams_ = csOne;
```

Comment:
There are actually four data members in CostMethodHbaseInsert. One of them is hidden in the code review tool because those few lines were unchanged. (In the old code, there was a CostMethodDP2Insert class, which I renamed CostMethodHbaseInsert. The four members in CostMethodHbaseInsert were all taken unchanged from CostMethodDP2Insert. Other members in CostMethodDP2Insert were deleted.)